app/supplier_utils.py

```diff
@@ -37,3 +44,80 @@ def check_supplier_role(role, supplier_id):
+        supplier_framework.application_company_details_confirmed
+    )
+
+
+def update_open_declarations_with_company_details(db, supplier_id, updater_json, specific_framework_slug=None):
+    """Expected to be called within a view"""
+    open_supplier_frameworks_query = SupplierFramework.query.filter(
+        SupplierFramework.supplier_id == supplier_id,
+    )
+
+    if specific_framework_slug:
+        open_supplier_frameworks_query = open_supplier_frameworks_query.filter(
+            SupplierFramework.framework.has(Framework.slug == specific_framework_slug),
+        )
+    else:
+        open_supplier_frameworks_query = open_supplier_frameworks_query.filter(
+            SupplierFramework.framework.has(Framework.status == 'open'),
+        )
+
+    # We need to ensure that SupplierFramework and Framework haven't changed during this transaction, so grab
+    # the affected rows with a for update lock.
+    open_supplier_frameworks = open_supplier_frameworks_query.options(lazyload('*')).with_for_update().all()
+
+    if not open_supplier_frameworks:
+        return
+
+    supplier = Supplier.query.filter(
+        Supplier.supplier_id == supplier_id
+    ).first_or_404()
+
+    # Update the open framework application(s) with latest company details.
+    for open_supplier_framework in open_supplier_frameworks:
+        company_details = {
+            supplier_constants.KEY_TRADING_NAME: supplier.name,
+            supplier_constants.KEY_REGISTERED_NAME: supplier.registered_name,
+            supplier_constants.KEY_REGISTRATION_NUMBER: (
+                supplier.companies_house_number or supplier.other_company_registration_number
+            ),
+            supplier_constants.KEY_DUNS_NUMBER: supplier.duns_number,
+            supplier_constants.KEY_VAT_NUMBER: supplier.vat_number,
+            supplier_constants.KEY_TRADING_STATUS: supplier.trading_status,
+            supplier_constants.KEY_ORGANISATION_SIZE: supplier.organisation_size,
+            supplier_constants.KEY_REGISTRATION_COUNTRY: supplier.registration_country,
+            supplier_constants.KEY_REGISTRATION_BUILDING: supplier.contact_information[0].address1,
+            supplier_constants.KEY_REGISTRATION_TOWN: supplier.contact_information[0].city,
+            supplier_constants.KEY_REGISTRATION_POSTCODE: supplier.contact_information[0].postcode,
+        }
+
+        company_details_without_nones = {k: v for k, v in company_details.items() if v is not None}
+
+        open_supplier_framework.declaration = {
+            **open_supplier_framework.declaration,
+            **company_details_without_nones
+        }
+
+        open_supplier_framework.company_details_confirmed = False
+
+        db.session.add(open_supplier_framework)
+
+        calling_function = inspect.stack()[1].function
+        db.session.add(
+            AuditEvent(
+                audit_type=AuditTypes.answer_selection_questions,
+                db_object=open_supplier_framework,
```

Comment:
Potentially the object should be the supplier for ease of search, as we can't currently search the data?

Reply:
The other place that does an `answer_selection_questions` audit event has the object as the supplier framework, so this is done for consistency's sake.